### PR TITLE
Comparing DB view with Avro file

### PR DESCRIPTION
**Pull request by dmorokov**
Comparing DB view with Avro file

1. Comparing functionality
2. Implementing SLF4J and SLF4J-JDK14
3. Adjusting Logger default message format
4. Adding JVM arg into Maven Surefire plugin

### CODE REVIEW

File: integ-tests/src/test/java/com/google/integration/IntegrationTests.java [hunk 1]
```diff
@@ -80,5 +86,31 @@ public void usersTest() throws SQLException, IOException {
...
+      logger.error("DB view and Avro file have mutually exclusive row(s)");
+
+      logger.info(format("DB view '%s' has %d different row(s)", sql, dbList.size()));
+      dbList.forEach(System.out::println);
+
+      logger.info(format("Avro file %s has %d different row(s)", avroFile, avroList.size()));
+      avroList.forEach(System.out::println);
+
+      Assert.fail();
+    } else if (dbList.size() > 0) {
```
Comment (haselbach) [inline integ-tests/src/test/java/com/google/integration/IntegrationTests.java:106]: Why not just use `!dbList.isEmpty()`?

File: integ-tests/src/test/java/com/google/integration/IntegrationTests.java [hunk 2]
```diff
@@ -80,5 +86,31 @@ public void usersTest() throws SQLException, IOException {
...
+    avroList.forEach(dbList::remove);
+    dbListCopy.forEach(avroList::remove);
+
+    if (dbList.size() == 0 && avroList.size() == 0) {
+      logger.info("DB view and Avro file are equal");
+    } else if (dbList.size() > 0 && avroList.size() > 0) {
+      logger.error("DB view and Avro file have mutually exclusive row(s)");
+
+      logger.info(format("DB view '%s' has %d different row(s)", sql, dbList.size()));
+      dbList.forEach(System.out::println);
```
Comment (haselbach) [inline integ-tests/src/test/java/com/google/integration/IntegrationTests.java:100]: Is this debugging output? If this is meant to stay, shouldn't this rather use the same logger as above?

Comment (dmorokov) [reply integ-tests/src/test/java/com/google/integration/IntegrationTests.java:100]: Yep, valid point, but there is a small inconvenience:
1. Logger standard message looks like -
Jan 10, 2022 7:22:05 PM com.google.integration.IntegrationTests usersTest
INFO: DB view 'src/main/java/com/google/sql/users.sql' has 16 different row(s)
2. Meanwhile println put everything line by line -
{UserName: {string: DBC}, CreatorName: {string: DBC}, CreateTimeStamp: {long: 1503419580000}, LastAccessTimeStamp: {long: 0}}
{UserName: {string: All}, CreatorName: {string: DBC}, CreateTimeStamp: {long: 1503419580000}, LastAccessTimeStamp: {long: 0}}

I can adjust the CLI output format for SLF4J but it will make a message really long.
If you are OK with this, then I'll do it.

File: integ-tests/src/main/java/com/google/pojo/UserTableRow.java [hunk 3]
```diff
@@ -25,29 +27,58 @@
...
+        && Objects.equals(lastAccessTimeStamp, that.lastAccessTimeStamp);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(userName, creatorName, createTimeStamp, lastAccessTimeStamp);
+  }
+
   /** Inner POJO class for serialization data from DB and Avro files. */
   public class UserName {
```
[the remark this answers is not in the document]
Comment (dmorokov) [reply integ-tests/src/main/java/com/google/pojo/UserTableRow.java:63]: I don't mind doing this but what's the reason? 
Could you briefly explain?
I thought then less things you keep static then better from a memory utilization perspective.

File: integ-tests/src/test/java/com/google/integration/IntegrationTests.java (hunk 1 above)
Comment (dmorokov) [reply integ-tests/src/test/java/com/google/integration/IntegrationTests.java:106]: Done.

File: integ-tests/src/main/java/com/google/pojo/UserTableRow.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (haselbach) [reply integ-tests/src/main/java/com/google/pojo/UserTableRow.java:63]: By having the class be non-static, the instance needs to keep a reference to the enclosing class instance. It can also only be instantiated from within an instance of the outer class.

Why would having an inner class be non-static be better for memory utilization?

File: integ-tests/src/test/java/com/google/integration/IntegrationTests.java (hunk 2 above)
Comment (haselbach) [reply integ-tests/src/test/java/com/google/integration/IntegrationTests.java:100]: Having a long message is fine.

File: integ-tests/src/main/java/com/google/pojo/UserTableRow.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmorokov) [reply integ-tests/src/main/java/com/google/pojo/UserTableRow.java]: Done.

File: integ-tests/src/test/java/com/google/integration/IntegrationTests.java [hunk 4]
```diff
@@ -80,5 +86,31 @@ public void usersTest() throws SQLException, IOException {
           AvroHelper.getLongNotNull(record, "LastAccessTimeStamp");
       avroList.add(avroRow);
     }
+
+    List<UserTableRow> dbListCopy = new ArrayList<>(dbList);
```
Comment (haselbach) [inline integ-tests/src/test/java/com/google/integration/IntegrationTests.java:90]: I'm just curious about this pattern. Usually, tests use the pattern where the assert something and provide an error message if the assertion fails. in [ExtractExecutorImplTest line 73](For example https://github.com/google/dwh-assessment-extraction-tool/blob/main/src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/ExtractExecutorImplTest.java#L73) we assert that the outcome is 0.

Here you rather have an if check, log separately (i.e., not use the underlying unit test framework) and then fail manually.

I think the whole if/else ladder here could be replaced with

```
assertThat(dbList).isEqualTo(avroList);
```

IIRC, the truth library's [`IterableSubject#isEqualTo`](https://truth.dev/api/latest/com/google/common/truth/IterableSubject#isEqualTo(java.lang.Object)) provides a diff if the equals test fails.

In short, what is the purpose of the manual logging instead of letting the unit testing framework take care of that?

Comment (dmorokov) [reply integ-tests/src/test/java/com/google/integration/IntegrationTests.java:90]: I agree, this if-else ladder is bulky, but it gives a clear answer what went wrong printing out the proper diff.

Ok, I tried this 'assertThat(dbList).isEqualTo(avroList);' , I personally barely may understand what is missing or where there is an extra row. I can show you it on a call.

Comment (dmorokov) [reply integ-tests/src/test/java/com/google/integration/IntegrationTests.java:93]: Done.

File: integ-tests/src/test/java/com/google/integration/IntegrationTests.java (hunk 2 above)
Comment (dmorokov) [reply integ-tests/src/test/java/com/google/integration/IntegrationTests.java]: Done.

**Review by haselbach — APPROVED** [review]
